google/cloud/bigtable/_read_rows.py

```diff
@@ -0,0 +1,683 @@
+    ) -> AsyncGenerator[Any, None]:
+        """
+        Helper function to yield items from a buffer as an async generator
+        """
+        while True:
+            item = await buffer.get()
+            if item is StopAsyncIteration:
+                return
+            if isinstance(item, Exception):
+                raise item
+            yield item
+            await asyncio.sleep(0)
+
+    async def _read_rows_retryable_attempt(
+        self,
+        gapic_fn: Callable[..., Awaitable[AsyncIterable[ReadRowsResponse]]],
+        buffer_size: int,
+        per_request_timeout: float,
+        operation_deadline: float,
+        total_row_limit: int,
+    ) -> AsyncGenerator[Row | RequestStats, None]:
+        """
+        Retryable wrapper for merge_rows. This function is called each time
+        a retry is attempted.
+
+        Some fresh state is created on each retry:
+          - grpc network stream
+          - buffer for the stream
+          - state machine to hold merge chunks received from stream
+        Some state is shared between retries:
+          - last_seen_row_key is used to ensure that
+            duplicate rows are not emitted
+          - request is stored and (optionally) modified on each retry
+        """
+        if self._last_emitted_row_key is not None:
+            # if this is a retry, try to trim down the request to avoid ones we've already processed
+            try:
+                self._request["rows"] = _ReadRowsOperation._revise_request_rowset(
+                    row_set=self._request.get("rows", None),
+                    last_seen_row_key=self._last_emitted_row_key,
+                )
+            except _RowSetComplete:
+                # if there are no rows left to process, we're done
+                return
+            # revise next request's row limit based on number emitted
+            if total_row_limit:
+                new_limit = total_row_limit - self._emit_count
+                if new_limit <= 0:
+                    return
+                else:
+                    self._request["rows_limit"] = new_limit
+        params_str = f'table_name={self._request.get("table_name", "")}'
+        if self._request.get("app_profile_id", None):
+            params_str = (
+                f'{params_str},app_profile_id={self._request.get("app_profile_id", "")}'
+            )
+        time_to_deadline = operation_deadline - time.monotonic()
+        gapic_timeout = min(time_to_deadline, per_request_timeout)
+        new_gapic_stream = await gapic_fn(
+            self._request,
+            timeout=gapic_timeout,
+            metadata=[("x-goog-request-params", params_str)],
+        )
+        buffer: asyncio.Queue[Row | RequestStats | Exception] = asyncio.Queue(
+            maxsize=buffer_size
+        )
+        buffer_task = asyncio.create_task(
+            self._generator_to_buffer(buffer, new_gapic_stream)
+        )
+        buffered_stream = self._buffer_to_generator(buffer)
```

Comment:
in the scenario of:

```
count = 0
async for row in  table.read_rows_stream():
    count ++
    if count > 5:
       break
```
What will abort the underlying grpc stream?

Reply:
Sorry, missed this.

When a close is called, a `GeneratorExit` exception will be thrown in the retryable attempt where it is halted (on the [yield line](https://github.com/googleapis/python-bigtable/blob/4535c174cbaf965532b860112845d1794c211baa/google/cloud/bigtable/_read_rows.py#L208)). The only reference to the grpc stream is a local variable within this function, so In general, it should usually be safe to let the garbage collector handle cleanup for things that weren't scheduled as a background task

But I'll add an extra explicit close to be safe, since the grpc streams are resource intensive and important to get right